### DISCUSSION TITLE
{ggplot2} theme_set() with {targets} and crew_controller_local()

### DISCUSSION THREAD

**Discussion by gregrs-uk** [wlandau/crew#144] [Q&A] [answered]
Thanks for creating this package, which I'm hoping to use alongside the excellent {targets} package to speed up some pipelines.

I'm having difficulty setting a global {ggplot2} theme when using `crew_controller_local`. I've attached a MWE `_targets.R` below. When the `controller` option is commented out, the theme is correctly set. When `crew_controller_local` is used, it's not.

I assume this might be something to do with the [`ggplot_global` environment](https://github.com/tidyverse/ggplot2/blob/main/R/ggplot-global.R) used within [`ggplot2::theme_set()`](https://github.com/tidyverse/ggplot2/blob/main/R/theme-current.R).

I tried setting `reset_globals = FALSE` in the `crew_controller_local` call but this didn't help. Do you have any suggestions that would allow this to work?

(By the way, I'm using the version from the current `main` branch on Github: [87f4824](https://github.com/wlandau/crew/commit/87f4824a1bbbbda88158b276a356b4a1f8715513).)

```
library(targets)
library(crew)
library(ggplot2)

tar_option_set(
  # controller = crew_controller_local()
)

theme_set(theme_dark())

list(
  tar_target(
    plot_file,
    ggsave(
      "plot.png",
      ggplot(iris, aes(Sepal.Width, Sepal.Length)) + geom_point()
    ),
    format = "file"
  )
)
```

**Comment by wlandau** [comment, marked as answer]
I think this is because `ggplot2::theme_set()` sets the theme in a local package environment inside `ggplot2` instead the environment that `targets` ships to the workers (`tar_option_get("envir")`, which is usually `.GlobalEnv`).

```r
> ggplot2::theme_set
function (new) 
{
    old <- ggplot_global$theme_current
    ggplot_global$theme_current <- new
    invisible(old)
}
```

On each worker, `ggplot2` is loaded from scratch and does not know the modifications to the `ggplot_global` object hidden inside the `ggplot2` package environment. As an alternative, you could encapsulate `theme_set()` inside a global function and then call it in each ggplot target.

```r
library(targets)
library(crew)
library(ggplot2)

tar_option_set(
   controller = crew_controller_local()
)

set_theme <- function() {
  theme_set(theme_dark())
}

list(
  tar_target(
    name = plot_file,
    command = {
      set_theme()
      ggsave(
        "plot.png",
        ggplot(iris, aes(Sepal.Width, Sepal.Length)) + geom_point()
      )
    },
    format = "file"
  )
)
```

If you have many such targets, then [`tarchetypes::tar_hook_before()`](https://docs.ropensci.org/tarchetypes/reference/tar_hook_before.html) could make this process convenient.

**Comment by gregrs-uk** [comment]
Brilliant, thanks. [`tarchetypes::tar_hook_before()`](https://docs.ropensci.org/tarchetypes/reference/tar_hook_before.html) has done the job nicely.
